src/backends/perf_pt/decode.c

```diff
@@ -35,13 +35,317 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
+#define _GNU_SOURCE
+
+#include <stdio.h>
 #include <intel-pt.h>
+#include <pt_cpu.h>
+#include <stdbool.h>
+#include <inttypes.h>
+#include <stdint.h>
+#include <link.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
 
-void
-perf_pt_start_decode(void *buf) {
-    // XXX This function body is only here to check for correct libipt linkage
-    // at this time.
-    (void) buf;
+#include "perf_pt_private.h"
+
+struct load_self_image_args {
+    struct pt_image *image;
+    int vdso_fd;
+    char *vdso_filename;
+};
+
+// Private prototypes.
+static int drain_events(struct pt_block_decoder *, int);
+static bool load_self_image(struct load_self_image_args *);
+static int load_self_image_cb(struct dl_phdr_info *, size_t, void *);
+
+// Public prototypes.
+void *perf_pt_start_decode(void *, uint64_t, int, char *, int *);
+bool perf_pt_next_block(struct pt_block_decoder *, int *, uint64_t *);
+void perf_pt_free_decoder(struct pt_block_decoder *);
+
+/*
+ * Start decoding a PT trace.
+ *
+ * Accepts a raw buffer `buf` of length `len`.
+ *
+ * `vdso_fd` is an open file descriptor for the filename `vdso_filename`. This
+ * is where the VDSO code will be written. libipt will read this file lazily,
+ * so it's up to the caller to make sure this file lives long enough for their
```

Comment:
Does "the caller" mean "us"?

Reply:
The caller of this function, which is "us" in a sense. This detail is not exposed to the consumers of our library.

Reply:
Right, so I think being clearer that it's "us" will help.

Reply:
Hrm. Actually, I think that's clear from the source. Look at the sole call site of this function:
https://github.com/softdevteam/hwtracer/blob/ffa6ee0a521aa5f201d446a2c4aff4a979b3e3cb/src/backends/perf_pt/mod.rs#L87-L99